File: generated_code/cpp/solution_2347_3.cpp
#include <algorithm>

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
private:
    // Structure to return both the sum and the count of nodes in a subtree
    struct SubtreeInfo {
        int sum;
        int count;
    };

    int matchingNodesCount = 0;

    // Post-order traversal to calculate sum and count bottom-up
    // Time Complexity: O(N) - visits each node exactly once
    // Space Complexity: O(H) - where H is the height of the tree (recursion stack)
    SubtreeInfo calculateSubtree(TreeNode* node) {
        if (!node) {
            return {0, 0};
        }

        // Divide: Process left and right subtrees
        SubtreeInfo leftInfo = calculateSubtree(node->left);
        SubtreeInfo rightInfo = calculateSubtree(node->right);

        // Conquer: Combine results for the current subtree
        int totalSum = leftInfo.sum + rightInfo.sum + node->val;
        int totalCount = leftInfo.count + rightInfo.count + 1;

        // Check if the current node meets the condition (integer division truncates towards zero/rounds down)
        if (totalSum / totalCount == node->val) {
            matchingNodesCount++;
        }

        return {totalSum, totalCount};
    }

public:
    int averageOfSubtree(TreeNode* root) {
        matchingNodesCount = 0;
        calculateSubtree(root);
        return matchingNodesCount;
    }
};

File: generated_code/python/solution_2347_3.py
from typing import Optional, Tuple

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def averageOfSubtree(self, root: TreeNode) -> int:
        """
        Algorithm: Post-order Traversal (DFS)
        
        To calculate the average of each subtree, we need two pieces of information 
        from the left and right child subtrees:
        1. The sum of all node values in that subtree.
        2. The count of all nodes in that subtree.
        
        Using a post-order traversal (bottom-up approach) allows us to compute these 
        metrics for the current node efficiently in O(1) time once its children are processed.
        
        Time Complexity: O(N) where N is the number of nodes in the binary tree, 
                         as we visit each node exactly once.
        Space Complexity: O(H) where H is the height of the tree, representing the 
                          maximum memory consumed by the recursion stack.
        """
        self.matching_nodes_count = 0
        
        def calculate_subtree_metrics(node: Optional[TreeNode]) -> Tuple[int, int]:
            if not node:
                # Base case: (subtree_sum, subtree_count)
                return 0, 0
            
            # Recurse on left and right subtrees
            left_sum, left_count = calculate_subtree_metrics(node.left)
            right_sum, right_count = calculate_subtree_metrics(node.right)
            
            # Aggregate values for the current subtree
            current_sum = left_sum + right_sum + node.val
            current_count = left_count + right_count + 1
            
            # Calculate the average rounded down to the nearest integer
            if current_sum // current_count == node.val:
                self.matching_nodes_count += 1
                
            return current_sum, current_count
        
        calculate_subtree_metrics(root)
        return self.matching_nodes_count